src/primitives/transaction.h

```diff
@@ -351,6 +351,10 @@ inline void UnserializeTransaction(TxType& tx, Stream& s) {
             const_cast<CTxWitness*>(&tx.witness)->vtxinwit.resize(tx.vin.size());
             const_cast<CTxWitness*>(&tx.witness)->vtxoutwit.resize(tx.vout.size());
             s >> tx.witness;
+            if (!tx.HasWitness()) {
+                /* It's illegal to encode witnesses when all witness stacks are empty. */
+                throw std::ios_base::failure("Superfluous witness record");
+            }
```

Comment:
I think we already check this condition in the witness serialization itself, but doesn't hurt here either.